Avoid constructing CSSStyleSheetObservableArray in the common case where adoptedStyleSheets is not used

cdumez opened the pull request:
#### 527e395dc858280dd73281765761c52677d1573c
<pre>
Avoid constructing CSSStyleSheetObservableArray in the common case where adoptedStyleSheets is not used
<a href="https://bugs.webkit.org/show_bug.cgi?id=253906">https://bugs.webkit.org/show_bug.cgi?id=253906</a>

Reviewed by Darin Adler.

TreeScope.adoptedStyleSheets is a new feature and not commonly used. The
TreeScope constructor currently spend time constructing a
CSSStyleSheetObservableArray and detaching for it.

We now construct the CSSStyleSheetObservableArray lazily to avoid work in the
constructor. We also drop the work in the destructor as it is no longer
necessary given that CSSStyleSheetObservableArray uses WeakPtrs.

* Source/WebCore/css/CSSStyleSheet.cpp:
(WebCore::CSSStyleSheet::removeAdoptingTreeScope):
* Source/WebCore/css/CSSStyleSheet.h:
* Source/WebCore/css/CSSStyleSheetObservableArray.cpp:
(WebCore::CSSStyleSheetObservableArray::deleteValueAt):
(WebCore::CSSStyleSheetObservableArray::setSheets):
(WebCore::CSSStyleSheetObservableArray::willRemoveSheet):
(WebCore::CSSStyleSheetObservableArray::willDestroyTreeScope): Deleted.
* Source/WebCore/css/CSSStyleSheetObservableArray.h:
* Source/WebCore/dom/TreeScope.cpp:
(WebCore::TreeScope::TreeScope):
(WebCore::TreeScope::ensureAdoptedStyleSheets):
(WebCore::TreeScope::adoptedStyleSheets const):
(WebCore::TreeScope::adoptedStyleSheetWrapper):
(WebCore::TreeScope::setAdoptedStyleSheets):
(WebCore::TreeScope::~TreeScope): Deleted.
* Source/WebCore/dom/TreeScope.h:

Canonical link: <a href="https://commits.webkit.org/261702@main">https://commits.webkit.org/261702@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8809fa7b7add2a0f4aa00e3374c3a722a1f552fd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112594 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21746 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1255 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4370 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121142 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116661 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23087 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12909 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/5532 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118362 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/17152 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100365 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105663 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/99105 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/908 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/46171 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14080 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/944 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/95268 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14762 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10315 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20101 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52949 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/8167 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16602 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4463 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/dom/TreeScope.cpp:
```diff
@@ -553,19 +548,33 @@ RadioButtonGroups& TreeScope::radioButtonGroups()
 {
-    return m_adoptedStyleSheets->sheets();
+    static NeverDestroyed<Vector<RefPtr<CSSStyleSheet>>> emptyStyleSheets;
+    return m_adoptedStyleSheets ? m_adoptedStyleSheets->sheets() : emptyStyleSheets.get();
```
If I was adding this function now I would return a `span<const RefPtr<CSSStyleSheet>>` rather than a `const&` or expose a `forEach` function.

cdumez replied:
Good idea, switching to a Span.

cdumez committed:
Rename FrameView to LocalFrameView
https://bugs.webkit.org/show_bug.cgi?id=253926

Reviewed by Darin Adler.

Rename FrameView to LocalFrameView, as we are moving towards site isolation.
The RemoteFrameView counterpart already exists and I'll rename
AbstractFrameView to FrameView in a follow-up patch.

* Source/WebCore/*
* Source/WebKit/*:
* Source/WebKitLegacy/*:

Canonical link: https://commits.webkit.org/261691@main

tadeuzagallo committed:
[WGSL] Add API to revert AST modifications
https://bugs.webkit.org/show_bug.cgi?id=253699
rdar://106548133

Reviewed by Myles C. Maxfield.

Since we might use the same AST for compiling multiple times (and potentially with
different layouts), we can't leave our modifications to the AST to be visible by
the next compilation. Instead of directly modifying the AST this patch adds an API
for replacing nodes which records all the nodes that have been replaced so we can
undo it once we are done compiling.

* Source/WebGPU/WGSL/MangleNames.cpp:
(WGSL::NameManglerVisitor::introduceVariable):
(WGSL::NameManglerVisitor::readVariable const):
* Source/WebGPU/WGSL/WGSL.cpp:
(WGSL::prepareImpl):
* Source/WebGPU/WGSL/WGSLShaderModule.h:
(WGSL::ShaderModule::replace):
(WGSL::ShaderModule::revertReplacements):

Canonical link: https://commits.webkit.org/261692@main

cdumez committed:
Fix data race found by TSan in AudioScheduledSourceNode
https://bugs.webkit.org/show_bug.cgi?id=253898

Reviewed by Jer Noble.

AudioScheduledSourceNode::m_playbackState was accessed from several threads
without synchronization. Use std::atomic<> to address the issue.

* Source/WebCore/Modules/webaudio/AudioScheduledSourceNode.h:

Canonical link: https://commits.webkit.org/261693@main

Commit:
Fix wasm name section parsing
rdar://106657580

Reviewed by Yusuke Suzuki.

Simple bug fix for wasm custom name section locals parsing.

* JSTests/wasm/stress/simple-inline-stacktrace-2.js:
(i.catch):
* JSTests/wasm/stress/simple-inline-stacktrace.js:
(i.catch):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmNameSectionParser.cpp:
(JSC::Wasm::NameSectionParser::parse):
* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::SectionParser::parseCustom):

Canonical link: https://commits.webkit.org/261694@main

robert-jenner committed:
10X tests need pixel tolerance adjustments for running in simulators on arm64 Macs
https://bugs.webkit.org/show_bug.cgi?id=253918
rdar://106725307

Unreviewed test gardening.

Adjusting pixel tolerances to account for failures seen after upgrading macos version on bots from Monterey to Ventura.

* LayoutTests/css3/blending/background-blend-mode-body-image.html:
* LayoutTests/css3/blending/background-blend-mode-body-transparent-color-and-image.html:
* LayoutTests/css3/blending/background-blend-mode-body-transparent-image.html:
* LayoutTests/css3/filters/effect-grayscale-hw.html:
* LayoutTests/fast/images/sprite-sheet-image-draw.html:
* LayoutTests/imported/w3c/web-platform-tests/compat/webkit-linear-gradient-line-right.html:
* LayoutTests/imported/w3c/web-platform-tests/compat/webkit-linear-gradient-line-top.html:
* LayoutTests/svg/css/invalid-color-cascade.svg:
* LayoutTests/svg/gradients/spreadMethodDiagonal3.svg:
* LayoutTests/svg/gradients/spreadMethodDiagonal4.svg:

Canonical link: https://commits.webkit.org/261695@main

csaavedra committed:
Zydis fails to build in Debug mode with -Werror=type-limits (GCC 12.2 / SCCACHE)
https://bugs.webkit.org/show_bug.cgi?id=252309

Reviewed by Yusuke Suzuki.

Suppress -Wtype-limits while this gets fixed in zydis upstream.

Canonical link: https://commits.webkit.org/261696@main

emw-apple committed:
Add OSXChromiumDriver and support extra browser arguments in run-benchmark
https://bugs.webkit.org/show_bug.cgi?id=253553
rdar://106389711

Reviewed by Stephanie Lewis.

Support running benchmarks against Chromium in run-benchmark. Chromium
is like Chrome, except that it doesn't use the
--enable-field-trial-config switch, because field trials are enabled by
default.

Additionally, allow extra arguments to be passed to the browser process
through run-benchmark, to facilitate customized testing.

* Tools/Scripts/webkitpy/benchmark_runner/browser_driver/osx_chrome_driver.py:
(OSXChromiumDriver): Added.

* Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py:
(config_argument_parser): Add benchmark_args argument.

Remaining changes are all plumbing benchmark_args through to the browser
driver.

* Tools/Scripts/webkitpy/benchmark_runner/benchmark_runner.py:
(BenchmarkRunner.__init__):
* Tools/Scripts/webkitpy/benchmark_runner/browser_driver/browser_driver.py:
(BrowserDriver.__init__):
* Tools/Scripts/webkitpy/benchmark_runner/browser_driver/browser_driver_factory.py:
(BrowserDriverFactory.create):
* Tools/Scripts/webkitpy/benchmark_runner/browser_driver/osx_browser_driver.py:
(OSXBrowserDriver._launch_process):
(run_benchmark_plan):
* Tools/Scripts/webkitpy/benchmark_runner/webserver_benchmark_runner.py:
(WebServerBenchmarkRunner.__init__):
* Tools/Scripts/webkitpy/browserperfdash/browserperfdash_unittest.py:
(BrowserPerfDashRunnerTest.test_can_construct_runner_object_minimum_parameters):
(FakeBrowserDriver.__init__):

Canonical link: https://commits.webkit.org/261697@main

ddegazio committed:
Skip more tests with large WASM memories on memory-limited platforms
https://bugs.webkit.org/show_bug.cgi?id=253893
rdar://106216202

Reviewed by Justin Michaud.

Skips WASM stress tests externref-result-tuple.js and only-referenced.js on
memory-limited platforms since they allocate huge (~2.5GB) WASM memories.

* JSTests/wasm/stress/externref-result-tuple.js:
* JSTests/wasm/stress/only-referenced.js:

Canonical link: https://commits.webkit.org/261698@main

kkinnunen-apple committed:
WebCore::IOSurface should not store a reference to a CGContext
https://bugs.webkit.org/show_bug.cgi?id=253698
rdar://106547690

Reviewed by Matt Woodrow.

An IOSurface might be written to and read from by multiple CGContexts
or even other contexts such as Metal (via WebGL / ANGLE).

An arbitrary context that has been used is not useful anymore in the
future, as the caller will not know the context state.

Due to these points, does not make sense to store a CGContext as a member in
WebCore::IOSurface.

The only real CG draw to an IOSurface happens via ImageBufferIOSurfaceBackend.
Store the context there, as that is the place to use the context.

Caching the context behind the interface creates source of bugs, as in
here fixed WebAutomationSession / ViewSnapshot where the callers
would inadvertedly create the contexts and leave them stored.

Remove invalid comment about CGIOSurfaceContextCreate affecting the
use count. The added test proves otherwise.

* Source/WebCore/platform/graphics/cg/IOSurfacePool.cpp:
(WebCore::IOSurfacePool::willAddSurface):
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::create):
(WebCore::ImageBufferIOSurfaceBackend::ImageBufferIOSurfaceBackend):
(WebCore::ImageBufferIOSurfaceBackend::~ImageBufferIOSurfaceBackend):
(WebCore::ImageBufferIOSurfaceBackend::context):
(WebCore::ImageBufferIOSurfaceBackend::ensurePlatformContext):
(WebCore::ImageBufferIOSurfaceBackend::copyNativeImage):
(WebCore::ImageBufferIOSurfaceBackend::copyNativeImageForDrawing):
(WebCore::ImageBufferIOSurfaceBackend::sinkIntoNativeImage):
(WebCore::ImageBufferIOSurfaceBackend::releaseGraphicsContext):
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
(WebCore::IOSurface::create):
(WebCore::IOSurface::createFromImage):
(WebCore::IOSurface::createImage):
(WebCore::IOSurface::sinkIntoImage):
(WebCore::IOSurface::createPlatformContext):
(WebCore::IOSurface::setOwnershipIdentity):
(WebCore::IOSurface::ensurePlatformContext): Deleted.
(WebCore::IOSurface::releasePlatformContext): Deleted.
* Source/WebKit/UIProcess/Automation/cocoa/WebAutomationSessionCocoa.mm:
(WebKit::WebAutomationSession::platformGetBase64EncodedPNGData):
* Source/WebKit/UIProcess/mac/ViewSnapshotStoreMac.mm:
(WebKit::ViewSnapshot::setVolatile):
(WebKit::ViewSnapshot::asImageForTesting):
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBackend.cpp:
(WebKit::ImageBufferShareableMappedIOSurfaceBackend::create):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/261699@main

shallawa committed:
[GPU Process] Have one copy of NativeImage when it is shared between WebProcess and GPUProcess
https://bugs.webkit.org/show_bug.cgi?id=253655
rdar://106505650

Reviewed by Simon Fraser.

Since WebProcess has to share the local PlatformImage with GPUProcess through
ShareableBitmap, we can get rid of the local PlatformImage and have both processes
use the ShareableBitmap.

For GPUProcess, we can ensure we decode the PlatformImage for sizeForDrawing
instead of decoding it for the full size.

Also this patch simplifies DecodingOptions and make have explicit DecodingMode
and sizeForDrawing. This will make DecodingOptions::isCompatibleWith() simpler
and can be used for synchronous and asynchronous decoding mode.

* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::decoding const):
* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::frameImageAtIndexCacheIfNeeded):
(WebCore::BitmapImage::draw):
(WebCore::BitmapImage::internalStartAnimation):
(WebCore::BitmapImage::decode):
(WebCore::BitmapImage::imageFrameAvailableAtIndex):
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/DecodingOptions.h:
(WebCore::DecodingOptions::DecodingOptions):
(WebCore::DecodingOptions::decodingMode const):
(WebCore::DecodingOptions::isAuto const):
(WebCore::DecodingOptions::isSynchronous const):
(WebCore::DecodingOptions::isAsynchronous const):
(WebCore::DecodingOptions::sizeForDrawing const):
(WebCore::DecodingOptions::hasFullSize const):
(WebCore::DecodingOptions::hasSizeForDrawing const):
(WebCore::DecodingOptions::isCompatibleWith const):
(WebCore::DecodingOptions::operator== const): Deleted.
(WebCore::DecodingOptions::isAsynchronousCompatibleWith const): Deleted.
(WebCore::DecodingOptions::maxDimension): Deleted.
(WebCore::DecodingOptions::has const): Deleted.
(WebCore::DecodingOptions::hasDecodingMode const): Deleted.
(WebCore::DecodingOptions::hasSize const): Deleted.
* Source/WebCore/platform/graphics/GraphicsContext.h:
(WebCore::GraphicsContext::preferredImageDecodingMode const):
* Source/WebCore/platform/graphics/ImageFrame.cpp:
(WebCore::ImageFrame::hasFullSizeNativeImage const):
(WebCore::ImageFrame::hasDecodedNativeImageCompatibleWithOptions const):
* Source/WebCore/platform/graphics/ImageSource.cpp:
(WebCore::ImageSource::requestFrameAsyncDecodingAtIndex):
(WebCore::ImageSource::frameAtIndexCacheIfNeeded):
(WebCore::ImageSource::frameIsBeingDecodedAndIsCompatibleWithOptionsAtIndex):
(WebCore::ImageSource::frameImageAtIndexCacheIfNeeded):
* Source/WebCore/platform/graphics/ImageSource.h:
* Source/WebCore/platform/graphics/NativeImage.cpp:
(WebCore::NativeImage::NativeImage):
(WebCore::NativeImage::setPlatformImage):
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
(WebCore::createImageSourceThumbnailOptions):
(WebCore::appendImageSourceOption):
(WebCore::imageSourceThumbnailOptions):
(WebCore::ImageDecoderCG::createFrameImageAtIndex):
(WebCore::createImageSourceAsyncOptions): Deleted.
(WebCore::imageSourceAsyncOptions): Deleted.
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::decodingModeForImageDraw const):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::preferredImageDecodingMode const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::recordNativeImageUse):

Canonical link: https://commits.webkit.org/261700@main

cdumez committed:
Rename AbstractFrameView to FrameView
https://bugs.webkit.org/show_bug.cgi?id=253968

Reviewed by Darin Adler.

Rename AbstractFrameView to FrameView, now that FrameView was renamed to
LocalFrameView.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/Frame.h:
* Source/WebCore/page/FrameView.h: Renamed from Source/WebCore/page/AbstractFrameView.h.
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::virtualView const):
* Source/WebCore/page/LocalFrame.h:
* Source/WebCore/page/LocalFrameView.cpp:
* Source/WebCore/page/LocalFrameView.h:
(isType):
* Source/WebCore/page/RemoteFrame.cpp:
(WebCore::RemoteFrame::virtualView const):
* Source/WebCore/page/RemoteFrame.h:
* Source/WebCore/page/RemoteFrameView.cpp:
(WebCore::RemoteFrameView::setFrameRect):
* Source/WebCore/page/RemoteFrameView.h:
(isType):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:

Canonical link: https://commits.webkit.org/261701@main

cdumez committed:
Avoid constructing CSSStyleSheetObservableArray in the common case where adoptedStyleSheets is not used
https://bugs.webkit.org/show_bug.cgi?id=253906

Reviewed by Darin Adler.

TreeScope.adoptedStyleSheets is a new feature and not commonly used. The
TreeScope constructor currently spend time constructing a
CSSStyleSheetObservableArray and detaching for it.

We now construct the CSSStyleSheetObservableArray lazily to avoid work in the
constructor. We also drop the work in the destructor as it is no longer
necessary given that CSSStyleSheetObservableArray uses WeakPtrs.

* Source/WebCore/css/CSSStyleSheet.cpp:
(WebCore::CSSStyleSheet::removeAdoptingTreeScope):
* Source/WebCore/css/CSSStyleSheet.h:
* Source/WebCore/css/CSSStyleSheetObservableArray.cpp:
(WebCore::CSSStyleSheetObservableArray::deleteValueAt):
(WebCore::CSSStyleSheetObservableArray::setSheets):
(WebCore::CSSStyleSheetObservableArray::willRemoveSheet):
(WebCore::CSSStyleSheetObservableArray::willDestroyTreeScope): Deleted.
* Source/WebCore/css/CSSStyleSheetObservableArray.h:
* Source/WebCore/dom/TreeScope.cpp:
(WebCore::TreeScope::TreeScope):
(WebCore::TreeScope::ensureAdoptedStyleSheets):
(WebCore::TreeScope::adoptedStyleSheets const):
(WebCore::TreeScope::adoptedStyleSheetWrapper):
(WebCore::TreeScope::setAdoptedStyleSheets):
(WebCore::TreeScope::~TreeScope): Deleted.
* Source/WebCore/dom/TreeScope.h:

Canonical link: https://commits.webkit.org/261702@main
